src/main/scala/ServerSetup.scala

```diff
@@ -1,28 +1,130 @@
+import ServerSetup.route
+import UrlDatabaseImplementation._
+import UserDatabaseImplementation._
+import UserRoute._
 import akka.actor.ActorSystem
 import akka.http.scaladsl.Http
-import akka.http.scaladsl.model.{ContentTypes, HttpEntity}
-import akka.http.scaladsl.server.Directives.{complete, pathEndOrSingleSlash}
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport
+import akka.http.scaladsl.model.{ContentTypes, HttpEntity, StatusCodes}
+import akka.http.scaladsl.server.Directives._
+import spray.json._
 
-object ServerSetup extends App{
+trait SetupActor {
+  implicit val system = ActorSystem("SetupActor")
+}
+
+object ServerSetup extends UrlJsonProtocol with SprayJsonSupport {
 
+  val urlRoute =
+    path("urls") {
+      post {
+        entity(as[Url]) { url =>
+          storeUrl(url.userId, url.timestamp, url.value)
+          complete(StatusCodes.OK)
+        }
+      } ~
+        get {
+          (path(IntNumber) | parameter(Symbol("urlid").as[Int])) { urlId =>
+            getUrlById(urlId) match {
+              case Some(url) => complete(
+                HttpEntity(
+                  ContentTypes.`application/json`,
+                  url.toJson.prettyPrint
+                )
+              )
+              case None => complete(StatusCodes.NotFound)
+            }
+          } ~
+            parameter(Symbol("userid").as[Int]) { userId =>
+              val urls = getAllUrlsForUser(userId)
+              if (urls.isEmpty)
+                complete(HttpEntity(
+                  ContentTypes.`text/html(UTF-8)`,
+                  """
+                    |<html>
+                    | <body>
+                    |   User with the given ID doesn't have any URLs stored in the browser's history
+                    | </body>
+                    |</html>
+                    |""".stripMargin
+                ))
+              else
+                complete(
+                  HttpEntity(
+                    ContentTypes.`application/json`,
+                    urls.toJson.prettyPrint
+                  ))
+            } ~
+            pathEndOrSingleSlash {
+              complete(
+                HttpEntity(
+                  ContentTypes.`application/json`,
+                  getAllUrls.toJson.prettyPrint
+                ))
+            }
+        }
+    }
 
-  implicit val system = ActorSystem("Setup")
+  val registerAndLoginRoute = {
+      (path("register") & post) {
+        entity(as[User]) { user =>
+          insertUser(user.username, registerHashPass(user.password))
+          complete(StatusCodes.OK)
```

Comment:
This complete should depend on the result of the `insertUser`. In the case when `insertUser` is failing, you're still completing with 200 OK, which should not be the case.